File: project/jimmer-sql/src/test/java/org/babyfish/jimmer/sql/model/calc/BookStoreNewestBooksResolver.java
package org.babyfish.jimmer.sql.model.calc;

import org.babyfish.jimmer.sql.JSqlClient;
import org.babyfish.jimmer.sql.TransientResolver;
import org.babyfish.jimmer.sql.ast.Expression;
import org.babyfish.jimmer.sql.ast.tuple.Tuple2;
import org.babyfish.jimmer.sql.model.BookStoreTable;
import org.babyfish.jimmer.sql.model.BookTable;

import java.util.*;
import java.util.stream.Collectors;

public class BookStoreNewestBooksResolver implements TransientResolver<UUID, List<UUID>> {

    private static final BookStoreTable table = BookStoreTable.$;

    private final JSqlClient sqlClient;

    public BookStoreNewestBooksResolver(JSqlClient sqlClient) {
        this.sqlClient = sqlClient;
    }

    @Override
    public Map<UUID, List<UUID>> resolve(Collection<UUID> ids) {
        BookTable book = new BookTable();
        List<Tuple2<UUID, UUID>> tuples = sqlClient.createQuery(table)
                .where(
                        Expression.tuple(
                                table.asTableEx().books().name(),
                                table.asTableEx().books().edition()
                        ).in(
                                sqlClient
                                        .createSubQuery(book)
                                        .where(book.store().id().in(ids))
                                        .groupBy(book.name())
                                        .select(
                                                book.name(),
                                                book.edition().max()
                                        )
                        )
                )
                .select(
                        table.id(),
                        table.asTableEx().books().id()
                )
                .execute(TransientResolver.currentConnection());
        return tuples.stream().collect(
                Collectors.groupingBy(
                        Tuple2<UUID, UUID>::get_1,
                        Collectors.mapping(
                                Tuple2<UUID, UUID>::get_2,
                                Collectors.toList()
                        )
                )
        );
    }
}

File: project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/model/calc/BookStoreNewestBooksResolver.kt
package org.babyfish.jimmer.sql.kt.model.calc

import org.babyfish.jimmer.sql.kt.KSqlClient
import org.babyfish.jimmer.sql.kt.KTransientResolver
import org.babyfish.jimmer.sql.kt.ast.expression.asNonNull
import org.babyfish.jimmer.sql.kt.ast.expression.max
import org.babyfish.jimmer.sql.kt.ast.expression.tuple
import org.babyfish.jimmer.sql.kt.ast.expression.valueIn
import org.babyfish.jimmer.sql.kt.model.*

class BookStoreNewestBooksResolver(
    private val sqlClient: KSqlClient
) : KTransientResolver<Long, List<Long>> {

    override fun resolve(ids: Collection<Long>): Map<Long, List<Long>> {
        val tuples = sqlClient.createQuery(BookStore::class) {
            val book = table.asTableEx().books
            where(
                tuple(book.name, book.edition) valueIn subQuery(Book::class) {
                    where(table.store.id valueIn ids)
                    groupBy(table.name)
                    select(table.name, max(table.edition).asNonNull())
                }
            )
            select(
                table.id,
                book.id
            )
        }.execute(KTransientResolver.currentConnection)
        return tuples.groupBy({it._1}) {
            it._2
        }
    }
}